src/plots/cartesian/align_period.js

```diff
@@ -0,0 +1,113 @@
+/**
+* Copyright 2012-2020, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+'use strict';
+
+var isNumeric = require('fast-isnumeric');
+var Lib = require('../../lib');
+var ms2DateTime = Lib.ms2DateTime;
+var dateTime2ms = Lib.dateTime2ms;
+var constants = require('../../constants/numerical');
+var ONEDAY = constants.ONEDAY;
+var ONEAVGMONTH = constants.ONEAVGMONTH;
+var ONEAVGYEAR = constants.ONEAVGYEAR;
+
+module.exports = function alignPeriod(trace, ax, axLetter, vals) {
+    if(ax.type !== 'date') return vals;
+
+    var alignment = trace[axLetter + 'periodalignment'];
+    if(!alignment) return vals;
+
+    var period = trace[axLetter + 'period'];
+    if(isNumeric(period)) {
+        period = +period;
+        if(period <= 0) return vals;
+    } else if(typeof period === 'string' && period.charAt(0) === 'M') {
+        var n = +(period.substring(1));
+        if(n > 0 && Math.round(n) === n) period = n * ONEAVGMONTH;
+        else return vals;
+    }
+
+    if(period > 0) {
+        var calendar = ax.calendar;
+
+        var isStart = 'start' === alignment;
+        // var isMiddle = 'middle' === alignment;
+        var isEnd = 'end' === alignment;
+
+        var offset = (new Date()).getTimezoneOffset() * 60000;
+        var period0 = trace[axLetter + 'period0'];
+        var base = (dateTime2ms(period0, calendar) || 0) - offset;
+
+        var newVals = [];
+        var len = vals.length;
+        for(var i = 0; i < len; i++) {
+            var v = vals[i] - base;
+
+            var dateStr = ms2DateTime(v, 0, calendar);
+            var O = new Date(dateStr);
```

Comment:
O or 0? 😅 this might go away anyway given my other [comment](https://github.com/plotly/plotly.js/pull/5074/files#r490279882) but let's not use O as a var name 🙏

Reply:
Done in https://github.com/plotly/plotly.js/pull/5074/commits/6b68e340b3181392454734e2880b7d98c7e60c3b.